## Normalisation

For simplicity, we will use the simplest method: library size normalisation and log transformation, but there are others
[@risso2018general, @hafemeister2019normalization]. 

Steps:

- Transcript length normalisation: 
    - Adjusts for differences in transcript length between genes
    - Divides the counts in each cell by the length of the transcript
    - Not necessary if you sequence a fixed region of the transcript ( 3' or 5' in 10X, our case)
- Library size normalisation: 
    - Adjusts for differences in sequencing depth between cells
    - Divides the counts in each cell by the total counts in that cell and multiplies by a scale factor (e.g. 10,000)

```{python}
sc.pp.normalize_total(adata, target_sum=1e4)
```

- Log transformation:
    - Logarithm of the normalised counts
    - Makes the data more normally distributed

```{python}
sc.pp.log1p(adata)
```