
## Set up 

:::{.panel-tabset}

### Google Colab

Go to [Google Colab](https://colab.research.google.com/)

Open a notebook.

![](figures/GoogleColab0.png)

Open an notebook from Github.

![](figures/GoogleColab1.png)

Introduce `https://github.com/IreneRobles/scrnaseq_workshops` and select `workshop_notebook.ipynb`

![](figures/GoogleColab2.png)

If you run notebook set up commands, it should install required packages and download test data.

### Conda and jupyter

Download repo

```bash
git clone https://github.com/IreneRobles/scrnaseq_workshops.git
cd /scrnaseq_workshops
```

Create environment

```bash
conda create -n myscanpy python=3.10
conda activate myscanpy
pip install -r requirements.txt
```

Register jupyter kernel

```bash
python -m ipykernel install --user --name myscanpy --display-name "Python (myscanpy)"
```
Download test data

```bash
mkdir data
cd data
wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE190nnn/GSE190622/suppl/GSE190622%5Fcount%5Fmatrix%5FAnnotated.csv.gz
```

:::

### Import libraries

```{python}
import scanpy as sc
import scrublet as scr
import numpy as np
import os
import pandas as pd
import polars as pl
import matplotlib.pyplot as plt
import seaborn as sns
```

Scanpy setttings
```{python}
sc.settings.verbosity = 3   # verbosity: errors (0), warnings (1), info (2), hints (3)
sc.logging.print_header()
sc.settings.set_figure_params(dpi=80, facecolor='white')
```

### Load tests datasets

Single-cell SMART-seq data from mouse macrophages [@robles2022cohesin]

```{python}
mf = pd.read_csv("data/GSE190622_count_matrix_Annotated.csv.gz", index_col=0)
adata = sc.AnnData(X=mf.T)
adata.obs["genotype"] = adata.obs_names.to_series().apply(lambda x: x.split("_")[0])
adata.obs["timepoint"] = adata.obs_names.to_series().apply(lambda x: x.split("_")[1])
adata.obs["sample"] = adata.obs["genotype"] + "_" + adata.obs["timepoint"]
adata.var_names_make_unique()
adata
```